File: backend/src/main/java/com/poeflipfinder/backend/gateway/CurrencyReferenceGateway.java
package com.poeflipfinder.backend.gateway;

import com.poeflipfinder.backend.entity.Currency;
import java.util.Optional;

/**
 * Defined by the use-case ring: resolve a Currency Exchange item path to a
 * persisted Currency row, creating one on first sight via ItemIconGateway.
 * Empty means the item genuinely can't be resolved even there -- the
 * interactor treats that as skip-this-pair, not a run failure (see
 * docs/ARCHITECTURE.md § Currency Exchange Ingestion).
 */
public interface CurrencyReferenceGateway {

    Optional<Currency> resolveOrCreateCurrency(String externalId);
}

File: app/lib/poe_flip_finder/currency_reference_gateway.ex
defmodule PoeFlipFinder.CurrencyReferenceGateway do
  @moduledoc """
  Defined by the core: resolve a Currency Exchange item path to a persisted
  Currency row, creating one on first sight via `PoeFlipFinder.ItemIconGateway`.
  `nil` means the item genuinely can't be resolved even there -- the caller
  treats that as skip-this-pair, not a run failure (see
  docs/ARCHITECTURE.md § Currency Exchange Ingestion).
  """

  alias PoeFlipFinder.Currency

  @callback resolve_or_create_currency(external_id :: String.t()) :: Currency.t() | nil
end
